test_runner/regress/test_remote_storage.py

```diff
@@ -100,45 +129,155 @@ def test_remote_storage_backup_and_restore(
+            assert (
+                query_scalar(
+                    cur,
+                    f"SELECT COUNT(*) FROM foo WHERE c = {checkpoint_id} AND t = '{FILLER_STRING}'",
+                )
+                == CHECKPOINT_ROWS
+            )
+
+
+def populate_tenant(
+    ps_http: NeonPageserverHttpClient,
+    cli: NeonCli,
+    postgres_factory: PostgresFactory,
+    tenant_id: TenantId,
+    timeline_id: TimelineId,
+    checkpoint_ids: List[int],
+):
+    cli.create_tenant(tenant_id=tenant_id, timeline_id=timeline_id)
+
+    # should get the same timeline id because it is the only one
+    pg = postgres_factory.create_start("main", tenant_id=tenant_id)
+
+    populate_checkpoints(
+        pg=pg,
+        ps=ps_http,
+        tenant_id=tenant_id,
+        timeline_id=timeline_id,
+        checkpoint_ids=checkpoint_ids,
+    )
+    return pg
+
+
+@pytest.mark.parametrize("remote_storage_kind", [RemoteStorageKind.LOCAL_FS])
+def test_remote_storage_corrupt_index_part(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    """
+    create two tenants
+    corrupt metadata in the remote index part json on one of them
+    check that pageserver loaded, and non corrupted timeline works as expected
+    """
+
+    # populate data
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_remote_storage_corrupt_index_part",
+    )
+
+    env = neon_env_builder.init_start()
+    ps_http = env.pageserver.http_client()
+
+    tenant_id1 = TenantId.generate()
+    timeline_id1 = TimelineId.generate()
+
+    checkpoint_ids = list(range(1, 3))
+
+    pg1 = populate_tenant(
+        ps_http=ps_http,
+        cli=env.neon_cli,
+        postgres_factory=env.postgres,
+        tenant_id=tenant_id1,
+        timeline_id=timeline_id1,
+        checkpoint_ids=checkpoint_ids,
+    )
+    pg1.stop()
+
+    tenant_id2 = TenantId.generate()
+    timeline_id2 = TimelineId.generate()
+
+    pg2 = populate_tenant(
+        ps_http=ps_http,
+        cli=env.neon_cli,
+        postgres_factory=env.postgres,
+        tenant_id=tenant_id2,
+        timeline_id=timeline_id2,
+        checkpoint_ids=checkpoint_ids,
+    )
+    pg2.stop()
+
+    # stop pageserver, corrupt metadata
+    env.pageserver.stop()
+
+    index_part = read_local_fs_index_part(env, tenant_id1, timeline_id1)
+    # corrupt metadata body, so it is not header but real body
+
+    # XXX: should we check for header corruptions too?
+    index_part["metadata_bytes"][16] += 1
+
+    write_local_fs_index_part(env, tenant_id1, timeline_id1, index_part)
+
+    env.pageserver.start()
+
+    # I would imagine that tenant with corrupted remote metadata wont be registered as healthy one
+    # but in fact because error occurs only during remote index build error does not get propagated to the
+    # registration code, so pageserver starts this timeline with all the background threads active
```

Comment:
Huh, yeah, that sounds bad

Reply:
I would imagine, https://github.com/neondatabase/neon/pull/2651 could help with this